packages/firebase_dynamic_links/android/src/main/java/io/flutter/plugins/firebasedynamiclinks/FirebaseDynamicLinksPlugin.java

```diff
@@ -0,0 +1,224 @@
+
+          result.success(url);
+        } else {
+          Exception exception = task.getException();
+          String errMsg = "Unable to create short link";
+          if (exception != null && exception.getLocalizedMessage() != null) {
+            errMsg = exception.getLocalizedMessage();
+          }
+          result.error("short_link_error", errMsg, null);
+        }
+      }
+    };
+  }
+
+  private void buildShortDynamicLink(
+      DynamicLink.Builder builder, MethodCall call, OnCompleteListener<ShortDynamicLink> listener) {
+    Integer suffix = null;
+
+    Map<String, Object> dynamicLinkParametersOptions =
+        call.argument("dynamicLinkParametersOptions");
+    if (dynamicLinkParametersOptions != null) {
+      Integer shortDynamicLinkPathLength =
+          (Integer) dynamicLinkParametersOptions.get("shortDynamicLinkPathLength");
+      if (shortDynamicLinkPathLength != null) {
+        switch (shortDynamicLinkPathLength) {
+          case 0:
+            suffix = ShortDynamicLink.Suffix.UNGUESSABLE;
+            break;
+          case 1:
+            suffix = ShortDynamicLink.Suffix.SHORT;
+            break;
+          default:
+            break;
+        }
+      }
+    }
+
+    if (suffix != null) {
+      builder.buildShortDynamicLink(suffix).addOnCompleteListener(listener);
+    } else {
+      builder.buildShortDynamicLink().addOnCompleteListener(listener);
+    }
+  }
+
+  private DynamicLink.Builder setupParameters(MethodCall call) {
+    DynamicLink.Builder dynamicLinkBuilder = FirebaseDynamicLinks.getInstance().createDynamicLink();
+
+    String domain = call.argument("domain");
+    String link = call.argument("link");
+
+    dynamicLinkBuilder.setDynamicLinkDomain(domain);
+    dynamicLinkBuilder.setLink(Uri.parse(link));
+
+    Map<String, Object> androidParameters = call.argument("androidParameters");
+    if (androidParameters != null) {
+      DynamicLink.AndroidParameters.Builder builder =
+          new DynamicLink.AndroidParameters.Builder((String) androidParameters.get("packageName"));
+
+      String fallbackUrl = (String) androidParameters.get("fallbackUrl");
+      Integer minimumVersion = (Integer) androidParameters.get("minimumVersion");
+
+      if (fallbackUrl != null) builder.setFallbackUrl(Uri.parse(fallbackUrl));
+      if (minimumVersion != null) builder.setMinimumVersion(minimumVersion);
+
+      dynamicLinkBuilder.setAndroidParameters(builder.build());
+    }
+
+    Map<String, Object> googleAnalyticsParameters = call.argument("googleAnalyticsParameters");
+    if (googleAnalyticsParameters != null) {
+      String campaign = (String) googleAnalyticsParameters.get("campaign");
+      String content = (String) googleAnalyticsParameters.get("content");
+      String medium = (String) googleAnalyticsParameters.get("medium");
+      String source = (String) googleAnalyticsParameters.get("source");
+      String term = (String) googleAnalyticsParameters.get("term");
```

Comment:
You could consider adding a local helper method a la `MethodCall.argument(String)`
```java
static <T> T valueFor(String key, Map<String, Object> map) {
  @SuppressWarnings("unchecked")
  T result = (T) map.get(key);
  return result;
}
```
Then you can avoid all the casts.